### PR TITLE
Introduce SMILES property

### PR DESCRIPTION
**Pull request by merkys**
This PR is meant to serve as a base for discussion around introducing SMILES property for structures in cheminformatics namespace. I have taken the liberty to copy-paste my earlier attempt to introduce SMILES into the main OPTIMADE specification (https://github.com/Materials-Consortia/OPTIMADE/pull/392). I am aware that for some of the points there is no consensus:

1. String or special SMILES data type (see https://github.com/Materials-Consortia/OPTIMADE/pull/436 for data type and https://github.com/Materials-Consortia/OPTIMADE/pull/533 for property-specific `SEARCH` operator which could in this case support SMARTS searches)
2. One-valued or list-valued (personally I see little use of lists as SMILES define `.` as separator of unconnected molecular entities)
3. OpenSMILES, [IUPAC SMILES+](https://github.com/IUPAC/IUPAC_SMILES_plus) or something else
4. [Quirós et al. 2018](https://doi.org/10.1186/s13321-018-0279-6) or no additional recommendations where SMILES is ambiguous.

Tagging the people who showed interest in SMILES in OPTIMADE: @ml-evs @vaitkus @alex-belozerov @sauliusg @JPBergsma @rartino. Please tag those who I forgot.

### CODE REVIEW

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  version: "0.1.0"
+  format: "1.2"
+  name: "_cheminfo_smiles"
+  label: "_cheminfo_smiles_structures"
+type:
+  - "string"
+  - "null"
+description: |-
+  SMILES (Simplified Molecular Input Line Entry System) representation of the structure.
+  Values MUST adhere to the OpenSMILES specification v1.0 (http://opensmiles.org/opensmiles.html).
```
Comment (ml-evs) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:16]: Consensus is that we should recommend our favourite specifications, rather than necessarily enforcing it.

We can also recommend that implementations announce which flavour of SMILES they are using as human-readable metadata

Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:16]: > We can also recommend that implementations announce which flavour of SMILES they are using as human-readable metadata

There was a suggestion of having an enumerator for most-widely-used values and value `other` with another field defined for free-form text.

Comment (ml-evs) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:16]: Also just remembered this point that either this field should be multivalued, or we have a proper recommendation for how to deal with multiple disconnected chemical subcomponents, or the case where the SMILES does not include all atoms in the "unit cell"

Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml]: > Also just remembered this point that either this field should be multivalued, or we have a proper recommendation for how to deal with multiple disconnected chemical subcomponents, or the case where the SMILES does not include all atoms in the "unit cell"

I have mentioned this issue in the initial PR message, but please correct me if you meant a different thing.

The issue of not representing all atoms in the structure (sort of "unit cell") is a good catch, though.

**Comment by merkys** [comment]
I have added property metadata fields to support the identification of the used SMILES flavor.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  description: "This dictionary contains the per-entry metadata for the _cheminfo_smiles field."
+  x-optimade-type: "dictionary"
+  type:
+    - "object"
+    - "null"
+  properties:
+    _cheminfo_standard_name:
+      description: |-
+        A name of the standard the given SMILES string conforms to.
+        Takes value from an enumerator containing: 'IUPAC SMILES+' (https://github.com/IUPAC/IUPAC_SMILES_plus), 'OpenSMILES' (http://opensmiles.org/opensmiles.html) and 'other' (none of the above).
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:34]:
```suggestion
        Takes value from an enumeration set containing 'IUPAC SMILES+' (https://github.com/IUPAC/IUPAC_SMILES_plus), 'OpenSMILES' (http://opensmiles.org/opensmiles.html) and 'other' (none of the above).
```

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      type:
+        - "string"
+      examples:
+        - "Daylight"
+        - "OpenBabel"
+      x-optimade-unit: "inapplicable"
+    _cheminfo_standard_version:
+      description: |-
+        The version of the SMILES standard given in '_cheminfo_standard_name'.
+        MUST NOT contain the initial 'v' or 'V'.
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:59]: I would probably remove this, since we are not sure what versioning schema/rules the external project (specification, program, etc.) is using.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        - "IUPAC SMILES+"
+        - "OpenSMILES"
+        - "other"
+      examples:
+        - "IUPAC SMILES+"
+        - "OpenSMILES"
+      x-optimade-unit: "inapplicable"
+    _cheminfo_standard_name_other:
+      description: |-
+        A name of the standard the given SMILES string conforms to, if different from the enumerator values for '_cheminfo_standard_name'.
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:48]: I would rephrase the description to mention that this field should be used only it the "other" enumeration value is specified in '_cheminfo_standard_name'.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
+$$schema: "https://schemas.optimade.org/meta/v1.2/optimade/property_definition"
+$id: "https://schemas.optimade.org/namespaces/cheminformatics/v0.1/properties/structures/_cheminfo_smiles"
+title: "SMILES (Simplified Molecular Input Line Entry Specification) representation of the structure"
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:3]:
```suggestion
title: "SMILES (Simplified Molecular Input Line Entry System) representation of the structure"
```

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  - "CN1C=NC2=C1C(=O)N(C(=O)N2C)C"
+x-optimade-metadata-definition:
+  title: "Metadata for the _cheminfo_smiles field"
+  description: "This dictionary contains the per-entry metadata for the _cheminfo_smiles field."
+  x-optimade-type: "dictionary"
+  type:
+    - "object"
+    - "null"
+  properties:
+    _cheminfo_standard_name:
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:31]: "Specification" sound more natural than "standard" when talking about SMILES types, and we seem to already be such phrasing in the text (e.g. "OpenSMILES specification"). However, I am not adamant about this needing to be renamed.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 2 above)
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:34]: We might also want to mention that the 'other' value should be accompanied by an entry in the '_cheminfo_standard_name_other' field.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 6 above)
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:31]:
```suggestion
    _cheminfo_smiles_standard_name:
```

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        - "string"
+      enum:
+        - "IUPAC SMILES+"
+        - "OpenSMILES"
+        - "other"
+      examples:
+        - "IUPAC SMILES+"
+        - "OpenSMILES"
+      x-optimade-unit: "inapplicable"
+    _cheminfo_standard_name_other:
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:46]:
```suggestion
    _cheminfo_smiles_standard_name_other:
```

Comment (vaitkus) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:46]: But maybe it would make sense to call this field something like "_cheminfo_smiles_standard_details" or "_cheminfo_smiles_standard_special_details" and allow to provide a greater variety of information here?

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 3 above)
Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:59]: Makes sense.

File: src/v0.1.0/entrytypes/structures.yaml [hunk 8]
```diff
@@ -23,6 +23,12 @@ properties:
       sortable: false
       query-support: "none"
       response-level: "yes"
+  _cheminfo_smiles:
+    $$inherit: "/properties/structures/_cheminfo_smiles"
+    x-optimade-implementation:
+      sortable: false
+      query-support: "none"
+      response-default: true
```
[the remark this answers is not in the document]
Comment (merkys) [reply src/v0.1.0/entrytypes/structures.yaml]: Added all the supported operators, marking as resolved.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 2 above)
Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml]: Added in faf592f. However, now I am beginning to think that there should be a way for the provider to say "specification is other than the enumerated ones, but I am not sure which". Maybe then `other` value without any `_cheminfo_specification_name_other` could be used to convey this meaning.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        - "null"
+      enum:
+        - "IUPAC SMILES+"
+        - "OpenSMILES"
+        - "other"
+      examples:
+        - "IUPAC SMILES+"
+        - "OpenSMILES"
+      x-optimade-unit: "inapplicable"
+    _cheminfo_specification_name_other:
```
Comment (vaitkus) [inline src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:52]: I am happy to go with the "name_other" approach, however, it might be better to take the CIF approach and name the field "_cheminfo_specification_special_details". We could still encourage for this field to be filled only if the specification name is set to `other`, however, databases could also convey some additional details about the SMILES (e.g. "OpenSMILES specification is followed, but stereochemistry is not provided)".

Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:52]: Personally I am not a big fan of `special_details`. I know that for people coming from CIF world this is intuitive, but for others (and for me sometimes) this looks as if free-form catch-all text field to fill with any details which do not fit anywhere else.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  format: "1.2"
+  name: "_cheminfo_smiles"
+  label: "_cheminfo_smiles_structures"
+type:
+  - "string"
+  - "null"
+description: |-
+  SMILES (Simplified Molecular Input Line Entry System) representation of the structure.
+  Providers MAY canonicalize (i.e., use rules to establish the stable order of atoms) the produced SMILES representations, but this is not mandatory and no particular set of rules is recommended.
+  Generally, providers SHOULD NOT change the representation more frequently than the structure itself is modified.
```
[the remark this answers is not in the document]
Comment (ml-evs) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:17]: Sounds much better, thanks

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 9 above)
Comment (vaitkus) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:52]: Could we then introduce a separate catch-all field for the SMILES specification details? A name may be useful (e.g. referring to "Daylight SMILES" or "Chemaxon SMILES"), however, there may be more details regarding the SMILES treatment that a database may want to convey. For example, a database may use OpenSMILES, but completely avoid stereochemistry descriptions. In the case of the COD, we mainly follow the OpenSMILES standard, however, there is also a paper we could refer to that describes the depiction of certain compounds (organometallics, etc.).

File: src/v0.1.0/entrytypes/structures.yaml [hunk 11]
```diff
@@ -21,6 +21,13 @@ properties:
       sortable: false
       query-support: "none"
       response-default: true
+  _cheminfo_smiles:
+    $$inherit: "/properties/structures/_cheminfo_smiles"
+    x-optimade-implementation:
+      sortable: false
+      query-support: "partial"
+      query-support-operators: ['=', '!=', 'CONTAINS', 'IS KNOWN', 'IS UNKNOWN']
```
[the remark this answers is not in the document]
Comment (vaitkus) [reply src/v0.1.0/entrytypes/structures.yaml:29]: Among similar lines, one may want to match all structures that matches a certain SMILES string expect for several specific compounds (all caffeine-like compound, but not caffeine). The same could most likely be expressed with a more intricate SMARTS string, but this might be an easier approach.

**Comment by merkys** [comment]
I have updated the proposal according to the consensus reached today in OPTIMADE discussion section on cheminformatics. Please review and let me know should I missed any of the points.

I took the liberty to retain the `_cheminfo_specification_version` from the original proposal (was not discussed today), feel free to oppose to that.

File: src/v0.1.0/entrytypes/structures.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (merkys) [reply src/v0.1.0/entrytypes/structures.yaml]: @ml-evs today no animosity towards having `!=` was expressed, can we resolve?

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 7 above)
Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml]: We do not have to include the property names in metadata property fields, they are tied to specific properties already. I would go for shorter names.

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 2 above)
Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml]: > However, now I am beginning to think that there should be a way for the provider to say "specification is other than the enumerated ones, but I am not sure which". Maybe then `other` value without any `_cheminfo_specification_name_other` could be used to convey this meaning.

Added a sentence on `null` value in 39f6e0ed1ceb86beebc948032d73c054431e6fc7.

@vaitkus can we resolve?

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 9 above)
Comment (merkys) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml:50]: Added in 5ee3bc3bd1c89dca4b3d33f0365e8986118bc047, can we resolve?

File: src/v0.1.0/properties/structures/_cheminfo_smiles.yaml (hunk 7 above)
Comment (vaitkus) [reply src/v0.1.0/properties/structures/_cheminfo_smiles.yaml]: Ok. But we should then still rename it to `_cheminfo_specification_name_other` to match `_cheminfo_specification_name`.

**Review by vaitkus — APPROVED** [review]

File: src/v0.1.0/entrytypes/structures.yaml [hunk 12]
```diff
@@ -21,6 +21,13 @@ properties:
       sortable: false
       query-support: "none"
       response-default: true
+  _cheminfo_smiles:
+    $$inherit: "/properties/structures/_cheminfo_smiles"
+    x-optimade-implementation:
+      sortable: false
+      query-support: "partial"
+      query-support-operators: ['=', '!=', 'IS KNOWN', 'IS UNKNOWN']
```
Comment (tomlunderwood) [inline src/v0.1.0/entrytypes/structures.yaml]: Given that (for now) `_cheminfo_smiles` is a string, should we make it clear in the `description` (or elsewhere) for this that `=` and `!=` are basic string equality operators, and that for a client query such as `_cheminfo_smiles = x` (where `x` is a SMILES string) to match the molecule represented by `x` the client should ensure that `x` is the same flavour of SMILES as used by the database provider. I don't think this is _necessary_ to do, but would be helpful information to add.

**Review by tomlunderwood — APPROVED** [review]
Looks good to me. I've added a comment suggesting a possible improvement to the description.

File: src/v0.1.0/entrytypes/structures.yaml (hunk 12 above)
Comment (merkys) [reply src/v0.1.0/entrytypes/structures.yaml]: Described in df9aeaba5942e28011f81d656da6495807f40eef. Can we resolve?

**Review by a-e-day — APPROVED** [review]
All looks good!
